src/main/java/net/minecraftforge/common/crafting/PartialNBTIngredient.java

```diff
@@ -0,0 +1,157 @@
+    }
+
+    @Override
+    public boolean test(@Nullable ItemStack input)
+    {
+        if (input == null)
+            return false;
+        return items.contains(input.getItem()) && predicate.matches(input.getShareTag());
+    }
+
+    @Override
+    public boolean isSimple()
+    {
+        return false;
+    }
+
+    @Override
+    public IIngredientSerializer<? extends Ingredient> getSerializer()
+    {
+        return Serializer.INSTANCE;
+    }
+
+    @Override
+    public JsonElement toJson()
+    {
+        JsonObject json = new JsonObject();
+        json.addProperty("type", CraftingHelper.getID(Serializer.INSTANCE).toString());
+        if (items.size() == 1)
+        {
+            json.addProperty("item", items.iterator().next().getRegistryName().toString());
+        }
+        else
+        {
+            JsonArray items = new JsonArray();
+            // ensure the order of items in the set is deterministic when saved to JSON
+            this.items.stream().map(Item::getRegistryName).sorted().forEach(name -> items.add(name.toString()));
+            json.add("items", items);
+        }
+        json.addProperty("nbt", nbt.toString());
+        return json;
+    }
+
+    public static class Serializer implements IIngredientSerializer<PartialNBTIngredient>
+    {
+        public static final Serializer INSTANCE = new Serializer();
+
+        @Override
+        public PartialNBTIngredient parse(JsonObject json)
+        {
+            // parse items
+            Set<Item> items;
+            if (json.has("item"))
+                items = Set.of(CraftingHelper.getItem(GsonHelper.getAsString(json, "item"), true));
+            else if (json.has("items"))
+            {
+                ImmutableSet.Builder<Item> builder = ImmutableSet.builder();
+                JsonArray itemArray = GsonHelper.getAsJsonArray(json, "items");
+                for (int i = 0; i < itemArray.size(); i++)
+                {
+                    builder.add(CraftingHelper.getItem(GsonHelper.convertToString(itemArray.get(i), "items[" + i + ']'), true));
+                }
+                items = builder.build();
+            }
+            else
+                throw new JsonSyntaxException("Must set either 'item' or 'items'");
+
+            // parse NBT
+            if (!json.has("nbt"))
+                throw new JsonSyntaxException("Missing nbt, expected to find a String or JsonObject");
+            CompoundTag nbt = CraftingHelper.getNBT(json.get("nbt"));
+
+            return new PartialNBTIngredient(items, nbt);
+        }
+
+        @Override
+        public PartialNBTIngredient parse(FriendlyByteBuf buffer)
+        {
+            Set<Item> items = Stream.generate(() -> buffer.readRegistryIdUnsafe(ForgeRegistries.ITEMS)).limit(buffer.readVarInt()).collect(Collectors.toSet());
+            CompoundTag nbt = buffer.readNbt();
+            return new PartialNBTIngredient(items, Objects.requireNonNull(nbt));
```

Comment:
This check is when reading from network. We never write a null tag to the network, but the way it's implemented means we can technically read a null one (the return is marked nullable, in this case only possible from corrupt network data I assume). The constructor being nullable is not a bad idea, but it would be in addition to this, not replacing it.

Reply:
The goal here was to error immediately, instead of erroring later. Which is typically when I use that method.

The constructor is non null by default and this method returns nullable so I wrap it. If a modder is ignoring nullability and passing potentially null that is a bug in their code